From ce7e053cb5d76dc7c654186a2cb20248f437d6f7 Mon Sep 17 00:00:00 2001
From: Orivej Desh <orivej@gmx.fr>
Date: Thu, 17 Dec 2020 01:28:00 +0000
Subject: [PATCH] Support opening links with external browser

viewer->cur_link was not being set since press_button_cb has been
disabled because webkit2gtk has dropped webkit_web_view_get_hit_test_result.
---
 src/plugins/fancy/fancy_viewer.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/plugins/fancy/fancy_viewer.c b/src/plugins/fancy/fancy_viewer.c
index e504d4ec2..e2e8937ed 100644
--- a/src/plugins/fancy/fancy_viewer.c
+++ b/src/plugins/fancy/fancy_viewer.c
@@ -406,6 +406,12 @@ navigation_policy_cb (WebKitWebView    *web_view,
 		      WebKitPolicyDecisionType policy_decision_type,
 		      FancyViewer		*viewer)
 {
+	if (policy_decision_type != WEBKIT_POLICY_DECISION_TYPE_NAVIGATION_ACTION)
+		return false;
+	
+	WebKitNavigationPolicyDecision *navigation_decision = WEBKIT_NAVIGATION_POLICY_DECISION(policy_decision);
+	WebKitNavigationAction *navigation_action = webkit_navigation_policy_decision_get_navigation_action(navigation_decision);
+	viewer->cur_link = webkit_uri_request_get_uri(webkit_navigation_action_get_request(navigation_action));
 
 	debug_print("navigation requested to %s\n", viewer->cur_link);
 
@@ -417,11 +423,15 @@ navigation_policy_cb (WebKitWebView    *web_view,
         } else if (!strncmp(viewer->cur_link, "file://", 7) || !strcmp(viewer->cur_link, "about:blank")) {
             debug_print("local navigation request ACCEPTED\n");
             webkit_policy_decision_use(policy_decision);
+        } else if (viewer->override_prefs_external) {
+            debug_print("remote navigation request OPENED\n");
+            open_uri(viewer->cur_link, prefs_common_get_uri_cmd());
+            webkit_policy_decision_ignore(policy_decision);
         } else if (viewer->override_prefs_remote_content) {
             debug_print("remote navigation request ACCEPTED\n");
             webkit_policy_decision_use(policy_decision);
         } else {
-            debug_print("remote avigation request IGNORED\n");
+            debug_print("remote navigation request IGNORED\n");
             fancy_show_notice(viewer, _("Remote content loading is disabled."));
             webkit_policy_decision_ignore(policy_decision);
         }
-- 
2.29.2

